### ISSUE TITLE
[Pdr bot, sim] Implement two-sided prediction into sim_engine, pdr bot 1, pdr bot 3

### ISSUE THREAD

**Issue by trizin** [oceanprotocol/pdr-backend#599]
### TODOs

Parent epic: [#680](https://github.com/oceanprotocol/pdr-backend/issues/680)  Two-sided prediction bots

Berkay:
- [x] First cut of this issue (classifier, bot 3): https://github.com/oceanprotocol/pdr-backend/pull/613

Trent:
- [x] Make Aimodel do classification (not regression), where classifier outputs probabilities of class 1 vs 2 (up vs down).
- [x] In sim_engine predictoor: model two-sided prediction
- [x] In sim_engine trader: use up/down confidence level to decide amt of $ to put in
- [x] In sim_engine trader: sell when predict down, then buy back
- [x] In sim_engine: add contour plots to visualize models, when 2 input vars (and hide plot when not 2 vars)
- [x] In predictoor_agent approach 1: stake up/down as 50/50
- [x] In predictoor_agent approach 3: stake up=prob_up*stake_amt, down=(1-prob_up)*stake_amt
- [x] rename predictoor approach 3 --> approach 2. Incl CLI
- [x] predictoor.md README updated:
  - [x] approach 3 -> 2 
  - [x] "sim engine does 2-sided predicting, and trading"
  - [x] "pdr bot need 2 txs because two-sided predicting; second tx must be a different wallet; therefore need PRIVATE_KEY2 (here's how to export it)"
- [x] create follow-up issue: k8s works w 2 keys. README updated. Issue: [#678](https://github.com/oceanprotocol/pdr-backend/issues/678)
- [x] create follow-up issue: blog post on two-sided prediction [#679](https://github.com/oceanprotocol/pdr-backend/issues/679)

## Background / Motivation

Currently, at each epoch, a predictoor bot picks a direction (eg "up") then stakes eg 100 OCEAN for that.

The idea here is: the predictoor bot calculates a probability of up (eg 60%) and down (100 - 60 = 40%), then stakes the 100 OCEAN pro-rata to each (60 OCEAN for up, 40 OCEAN for down).

**Expected benefits:**
- reduces risk - because you'll always be catching some of the slashing
- reduces variance - because you're staking pro-rata to probabilities on both sides, rather than one.
  - this in turn probably increases the average expected return per epoch
  - this should work especially well when probabilities are close, eg 60-40 (vs 95-5).
- [Sim results](https://github.com/oceanprotocol/pdr-backend/issues/661#issuecomment-1959933240) show that one-sided prediction (status quo) is not competitive.

## Candidate implementations

**Cand 1: pdr-backend manages 2 private keys.**
- Pros: Simple mental model, logically cohesive. Both wallets managed from same place, only one instance is required. 
- Cons: Additional complexity in pdr-backend, to get Web3PP to hold >1 private key.

**Cand 2: envvars trick**
- What: Envvar: UP_STAKER, DOWN_STAKER, agents with one of these env vars only stake in that direction
- Pros: Low-ish complexity in mechanics of staking
- Cons: Messy mental model: the "amount to stake" bot logic needs to be unified across up & down. Leaves more work to users, min 2 instances are required. Testing is more complex.

**Cand 3: change smart contracts so that 1 wallet does up & down*
- What: new DT template (4), where a single wallet can stake up _and_ down
- Pros: most natural; pdr-backend doesn't need large workaround code 
- Cons: Additional complexity in smart contract; need to re-deploy contracts & manage community

## Discussion of candidates

Trent: 
- (3) is cleanest conceptually, but largest effort. 
- (1) and (2) are par for complexity add: (1) adds it in managing >1 wallet at once, (2) adds it in managing >1 instance at once. But the complexity-add of (1) is nicely abstracted away, inside Web3PP, and can be well tested. Whereas for (2) it will be harder to test, and will always be affecting the users
- Therefore I recommend (1).

### Related

- In [#661](https://github.com/oceanprotocol/pdr-backend/issues/661) "[Sim] Log & plot predicting $ made", we added logs & plots for profitability of predictoors. It's clear that one-sided prediction is not competitive.

**Comment by trentmc** [comment]
> @trentmc +1 to Web3PP solution. Should two wallets per predictoor become a requirement for both approach 1 and approach 3 or be optional? 

Let's keep approach 1 and 3 as they are, and make this approach 4. We'll keep approach 3 as the default because it's simpler onboarding.

> Additionally, could you point me to the piece of code where I can grab the accuracy

Approach 1 does no modeling. 

Approach 3 builds a _regression_ model (continuous-valued prediction). From that, it predicts a real-valued price. And then it converts to binary via `predval = predprice > curprice`. It's slightly more involved to get a probability from its setup. Therefore for approach 4...

Approach 4 will build a _classification_ model (binary prediction). From that, the model can give a probability. Let's use the "Logistic Regression" approach (which btw is a misnomer: despite "regression" in the label it's actually classification). Use its [`predict_proba()`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression.predict_proba) method to get probabilities.

`AimodelFactory` is tuned for regression. So let's bypass it. And for now, let's just build a classifier directly in the approach 4 code. Therefore here's what the code could look like (note I did not test this):
```python
from sklearn import linear_model
...

    def get_prediction_stakes(
        self, timestamp: int  # pylint: disable=unused-argument
    ) -> Tuple[bool, float]:
        """
        @description
          Give a stake for up prediction, and down prediction.

        @arguments
          timestamp -- int -- when to make prediction for (unix time)

        @return
          stake_up -- int -- amount to stake *up*, in units of Eth
          stake_down -- int -- amount to stake *down*, in units of Eth
        """
        mergedohlcv_df = self.get_data_components()

        model_data_factory = AimodelDataFactory(self.ppss.predictoor_ss)
        X, y, _ = model_data_factory.create_xy(mergedohlcv_df, testshift=0)

        # Split X/y into train & test data
        st, fin = 0, X.shape[0] - 1
        X_train, X_test = X[st:fin, :], X[fin : fin + 1]
        y_train_price, y_test_price = y[st:fin], y[fin : fin + 1] # continuous values
        curprice = y_test_price[-1]
        y_train_class = y_train_price > curprice # binary values

        # Compute classifier model 
        model = linear_model.LogisticRegression()
        model.fit(X_train, y_train_class)

        # Predict from test data
        y_test_class = model.predict_proba(X_test)
        prob_up = y_test_class[1][0] # or [0][0] ? 
        prob_down = 1.0 - prob_up

        # Stake amount
        stake = self.ppss.predictoor_ss.stake_amount
        stake_up = prob_up * stake
        stake_down = prob_down * stake

        return (stake_up, stake_down)
```

A couple things to check for:
- `curprice = y_test_price[-1]` may be wrong. See [#605](https://github.com/oceanprotocol/pdr-backend/issues/605) for a discussion on the continuous-valued side.
- `prob_up = y_test_class[1][0]` may actually need `[0][0]`. You'll have to play with it and see.

Finally, note that I've given this as `get_prediction_stakes()` returning `(stake_up, stake_down)`. Whereas for approach 1 and 3 it's `get_prediction()` returning `(up?, stake)`. You'll have to refactor base predictoor code (or other) accordingly. If you see a clean unified way for approach 1, 3, 4, where 1 & 3 don't need two keys, go for it!
